src/config.c

```diff
@@ -260,6 +293,115 @@ void configureVifs(void) {
     }
 }
 
+/* create VIFs for all IP, non-loop interfaces.
+   When argument is not NULL rebuild the interface table.
+*/
+void createVifs(struct IfDescP *RebuildP) {
+    struct IfDesc *Dp, *oDp = NULL;
+    int    vifcount = 0, upsvifcount = 0, Ix = 0;
+    bool   join = false;
+
+    // init array to "not set"
+    for (Ix = 0; Ix < MAX_UPS_VIFS; Ix++) {
+        upStreamIfIdx[Ix] = -1;
+    }
+
+    if (RebuildP != NULL) {
+        // When rebuild, check if interfaces have dissapeared and call delVIF if necessary.
+        for (oDp=RebuildP->S; oDp<RebuildP->E; oDp++) {
+            for (Ix = 0; (Dp = getIfByIx(Ix)); Ix++) {
+                if (! strcmp (oDp->Name, Dp->Name)) {
+                    break;
+                }
+            }
+            if (Dp == NULL) {
+                my_log(LOG_DEBUG, 0, "Interface %s disappeared from system", oDp->Name);
+                if (oDp->index != -1) {
+                    delVIF(oDp);
+                }
+            }
+        }
+    }
+
+    for(Ix = 0; (Dp = getIfByIx(Ix)); Ix++) {
+        if (RebuildP == NULL) {
+            // Only add vif for valid interfaces on start-up.
+            if ((Dp->Flags & IFF_LOOPBACK) || (Dp->state != IF_STATE_DOWNSTREAM && Dp->state != IF_STATE_UPSTREAM)) {
+                continue;
+            }
+        } else {
+            /* Need rebuild, check if interface is new or already exists (check table below).
+                             old: disabled    new: disabled    -> do nothing
+                             old: disabled    new: downstream  -> addVIF(new), joinmcroutergroups
```

Comment:
This calls `joinMcRoutersGroup` function which uses `setsockopt`/`IP_ADD_MEMBERSHIP` API to join mutlicast groups 224.0.0.2 (INADDR_ALLRTRS_GROUP) and 224.0.0.22 (INADDR_ALLIGMPV3_GROUP).

But it is doing it on downstream interface.

So what is purpose of this call? And why it is needed?

Reply:
This has been split off from initRoutetable, which is run upon startup to:
   // If this is a downstream vif, we should join the All routers group...

So if you have interfaces transitioning from disabled to downstream, these groups need to be joined.

Reply:
But why it is needed to be joined? For me it does not make sense to join these two particular multicast groups on downstream interface.

Downstream interface just needs to listen to IGMP packets (not whole multicast traffic sent to 224.0.0.2 / 224.0.0.22) and this is achieved by calling `MRT_INIT` on `AF_INET`/`SOCK_RAW`/`IPPROTO_IGMP` socket. This socket receives all IGMP packets (with and also without Router Alter bit).

Morover on downstream interface igmpproxy needs to listen for igmppackets not only sent to 224.0.0.2 or 224.0.0.22 multicast groups, but needs to listen for **all** igmp packet sent to any multicast group. But above `MRT_INIT` handles it.

So for me joining just two particular groups does not make sense (it it not enough) and it is not needed for igmp packets as there is already `MRT_INIT`.

Reply:
IGMP does indeed not require those groups to be joined. This is for consistency in behaviour. According to the documentation IP_ADD_MEMBERSHIP should be set per interface, also delVif() is accounting for leaving these groups. That may lead to situation where only 1 interface is downstream and this being removed and replaced by another the groups will no longer be joined if they are not joined on the new interface. Hence I opted to implement this behaviour. IGMP wise this may be removed completely, as there is no requirement, nor need for the proxy to join.

Reply:
I think new `joinMcRoutersGroup` function can be removed as I do not see reason for it. Calling it from this code is not needed and when changing this part of code it is better to "remove old unneeded code" as trying to reimplement it.

Reply:
Ok I removed the functionality to join mcrouters groups completely. Because initRoutetable() was left with only initializing *routetable to NULL I moved initialization to global declration, removed initRoutetable() and use clearRoutes() in igmpproxyinit() for good order.